docs/cudf/source/internals.md

```diff
@@ -0,0 +1,411 @@
+* `children`: A tuple of "children" Columns.
+  Children columns enable complex column types such as
+  Columns with non-fixed width elements (e.g., strings).
+
+Columns are specialized based on the type of data they contain:
+
+```cython
+# Column class hierarchy
+
+cdef class Column  # extension type
+
+class ColumnBase(Column)  # Python class
+
+class NumericalColumn(ColumnBase)
+class DatetimeColumn(ColumnBase)
+class StringColumn(ColumnBase)
+class CategoricalColumn(ColumnBase)
+```
+
+Note that `Column` is an
+[extension type](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html)
+defined using `cdef class`,
+a special type that can directly interface with C/C++ code.
+
+### Table and Frame
+
+A `Table` is a collection of Columns, with an optional `index` (itself a `Table`).
+
+```cython
+cdef class Table:
+    def __init__(self, data=None, index=None):
+        pass
+```
+
+* `data`: An `OrderedDict` of Columns
+* `index`: A `Table` representing the index of this Table
+
+```cython
+# Table class hierarchy
+
+cdef class Table  # extension type
+
+class Frame(Table)  # Python class
+
+class Index(Frame)
+class Series(Frame)
+class DataFrame(Frame)
+```
+
+Like `Column`, `Table` is an extension type.
+
+## Interfacing with libcudf <a name="interfacing"></a>
+
+`libcudf` is a C++/CUDA library that provides
+GPU accelerated implementations of common data operations
+such as sort, filter and groupby.
+Most operations on cuDF objects like DataFrames or Series
+are performed by libcudf under the hood.
+
+### Why Cython
+
+cuDF uses [Cython](https://cython.org) to interface with libcudf.
+While there are other options for using C/C++ code from Python,
+we use Cython becuase:
+
+1. It is used by other libraries in the PyData ecosystem such as
+   Pandas and NumPy.
+2. It supports  *modern* C++ features
+   such as smart pointers and move semantics.
+3. Cython code is much more similar to Python than it is to C/C++,
+   making it easier to use by Python programmers.
+
+### Wrapping a libcudf API - example
+
+This is the sort of thing best explained using an example,
+so we consider wrapping the libcudf function `gather`
+and exposing it in Python:
+
+```c++
+copying.pxd
+
+std::unique_ptr<cudf::table> gather(
+    cudf::table_view const& source_table,
+    cudf::column_view const& gather_map
+);
+```
+
+The `gather` operation "gathers" specified rows of a table,
+returning a new table that is composed of those rows.
+In cuDF, this operation is known as `take`:
+
+```python
+In [3]: a = cudf.DataFrame({'a': [1, 2, 3, 4], 'b': ['a', 'b', 'c', 'd']}, index=['x', 'y', 'z', 'w'])
+
+In [4]: a
+Out[4]:
+   a  b
+x  1  a
+y  2  b
+z  3  c
+w  4  d
+
+In [5]: a.take([0, 1, 3])
+Out[5]:
+   a  b
+x  1  a
+y  2  b
+w  4  d
+```
+
+A detailed overview of the libcudf library and explanation of the types
+`cudf::table`, `cudf::table_view`, etc.,
+can be found [here](https://github.com/rapidsai/rmm/)
```

Comment:
```suggestion
can be found [here](https://github.com/rapidsai/cudf/)
```

Reply:
Oops - I meant to link https://docs.google.com/presentation/d/1zKzAtc1AWFKfMhiUlV5yRZxSiPLwsObxMlWRWz_f5hA/